File: frontend/src/types/dss.ts
export interface CriterionDetail {
  raw_value: number;
  utility: number;
  weight: number;
  weighted_contribution: number;
  criterion_type: string;
}

export interface DSSRankingItem {
  rank: number;
  symbol: string;
  smart_score: number;
  criteria: Record<string, CriterionDetail>;
  reasons: string[];
}

export interface DSSRankingResponse {
  as_of_date: string;
  methodology: string;
  weights: Record<string, number>;
  rankings: DSSRankingItem[];
}

export interface SensitivityComparisonRow {
  symbol: string;
  balanced_score: number;
  balanced_rank: number;
  growth_score: number;
  growth_rank: number;
  growth_rank_change: number;
  risk_averse_score: number;
  risk_averse_rank: number;
  risk_averse_rank_change: number;
}

export interface SensitivityStabilitySummary {
  top_etf_stable: boolean;
  max_absolute_rank_change: number;
  growth_changed_count: number;
  risk_averse_changed_count: number;
}

export interface DSSSensitivityResponse {
  as_of_date: string;
  profiles: Record<string, Record<string, number>>;
  comparisons: SensitivityComparisonRow[];
  stability_summary: SensitivityStabilitySummary;
}

File: backend/app/schemas/dss.py
from datetime import date

from pydantic import BaseModel


class CriterionDetail(BaseModel):
    raw_value: float
    utility: float
    weight: float
    weighted_contribution: float
    criterion_type: str


class DSSRankingItem(BaseModel):
    rank: int
    symbol: str
    smart_score: float
    criteria: dict[str, CriterionDetail]
    reasons: list[str]


class DSSRankingResponse(BaseModel):
    as_of_date: date
    methodology: str
    weights: dict[str, float]
    rankings: list[DSSRankingItem]


class SensitivityComparisonRow(BaseModel):
    symbol: str
    balanced_score: float
    balanced_rank: int
    growth_score: float
    growth_rank: int
    growth_rank_change: int
    risk_averse_score: float
    risk_averse_rank: int
    risk_averse_rank_change: int


class SensitivityStabilitySummary(BaseModel):
    top_etf_stable: bool
    max_absolute_rank_change: int
    growth_changed_count: int
    risk_averse_changed_count: int


class DSSSensitivityResponse(BaseModel):
    as_of_date: date
    profiles: dict[str, dict[str, float]]
    comparisons: list[SensitivityComparisonRow]
    stability_summary: SensitivityStabilitySummary
